plugins/code-climate/src/api/production-api.ts

```diff
@@ -0,0 +1,184 @@
+      (duplicationResponse.status >= 400 && duplicationResponse.status < 600) ||
+      (otherIssuesResponse.status >= 400 && otherIssuesResponse.status < 600)
+    ) {
+      throw new Error('Failed fetching Code Climate info');
+    }
+
+    const maintainabilityData = (
+      (await maintainabilityResponse.json()) as CodeClimateMaintainabilityData
+    ).data.attributes.ratings[0];
+    const testCoverageData = (
+      (await testCoverageResponse.json()) as CodeClimateTestCoverageData
+    ).data.attributes.rating;
+    const codeSmellsData = (
+      (await codeSmellsResponse.json()) as CodeClimateIssuesData
+    ).meta.total_count;
+    const duplicationData = (
+      (await duplicationResponse.json()) as CodeClimateIssuesData
+    ).meta.total_count;
+    const otherIssuesData = (
+      (await otherIssuesResponse.json()) as CodeClimateIssuesData
+    ).meta.total_count;
+
+    return [
+      maintainabilityData,
+      testCoverageData,
+      codeSmellsData,
+      duplicationData,
+      otherIssuesData,
+    ];
+  }
+
+  async fetchData(repoID: string): Promise<CodeClimateData | null> {
+    if (!repoID) {
+      return null;
+    }
+
+    const apiUrl = `${await this.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/codeclimate/api`;
+    const options = await this.authOptions();
+
+    const repoResponse = await fetch(`${apiUrl}/repos/${repoID}`, options);
+
+    if (repoResponse.status >= 400 && repoResponse.status < 600) {
+      throw new Error('Failed fetching Code Climate info');
+    }
+
+    const repoData = ((await repoResponse.json()) as CodeClimateRepoData).data;
+
+    const [
+      maintainabilityData,
+      testCoverageData,
+      codeSmellsData,
+      duplicationData,
+      otherIssuesData,
+    ] = await this.fetchAllData(
+      apiUrl,
+      options,
+      repoID,
+      repoData.relationships.latest_default_branch_snapshot.data.id,
+      repoData.relationships.latest_default_branch_test_report.data.id,
+    );
+
+    const maintainabilityValue: any = {};
+    maintainabilityValue[
+      maintainabilityData.measure.meta.implementation_time.unit
+    ] = maintainabilityData.measure.meta.implementation_time.value.toFixed();
+
+    return {
+      repoID,
+      maintainability: {
+        letter: maintainabilityData.letter,
+        value: humanizeDuration(
+          Duration.fromObject(maintainabilityValue).toMillis(),
+          { largest: 1 },
+        ),
+      },
+      testCoverage: {
+        letter: testCoverageData.letter,
+        value: testCoverageData.measure.value.toFixed(),
+      },
+      numberOfCodeSmells: codeSmellsData,
+      numberOfDuplication: duplicationData,
+      numberOfOtherIssues: otherIssuesData,
+    };
+  }
+
+  private async authOptions() {
+    if (!this.identityApi) {
+      return {};
+    }
+    const { token } = await this.identityApi.getCredentials();
+    return {
+      headers: {
+        authorization: `Token token=${token}`,
```

Comment:
Ah this was meant for auth to the Backstage proxy itself rather than the underlying code climate service, is that right? Then this is actually a `Bearer <token>` auth, not `Token token=<token>`. And besides, the token can actually be undefined here.

But more importantly, instead of doing this yourself, you should actually stop requiring the `identityApi` and instead requiring (not even optionally!) the `fetchApi` instead. It will magically handle all of this auth concern stuff for you so you can delete all of this code and instead use a `this.fetchApi.fetch(...)` everywhere above. This may also make it unnecessary to support the `options` parameter in these methods entirely, right?